TensorLib/TensorData.lean

```diff
@@ -0,0 +1,226 @@
+-- within a section
+variable (a: Type)
+
+inductive TensorData where
+| td_root (xs: List a)
+| td_node (xs: NonEmptyList TensorData)
+
+instance : Inhabited (TensorData a) where
+  default := .td_root []
+
+def td_node! (xs: List (TensorData a)): TensorData a :=
+  .td_node (NonEmptyList.fromList! xs)
+
+-- theorem tensorDataInduction (P : TensorData a -> Prop)
+--  (hroot : ∀ (xs: List a), P (.td_root xs))
+--  (hnode : ∀ (xs: List (TensorData _)), (∀ x, x ∈ xs -> P x) -> P (.td_node xs))
+--  (x: TensorData a): P x :=
+--  match x with
+--   | .td_root xs => hroot xs
+--   | .td_node xs => hnode xs (fun x _ => tensorDataInduction P hroot hnode x)
+
+abbrev Shape := List Nat
+
+-- TODO: Make tail-recursive, perhaps when Lean will eliminite tail calls in mutual-recursion
+def TensorData.hasShape (x: TensorData a) (shape: Shape): Bool :=
+  match x, shape with
+  | .td_root y, List.cons n .nil => List.length y == n
+  | .td_node ys, List.cons n shape' =>
+      ys.length == n &&
+      -- `attach` is used in proofs below, e.g. HashShapeOk
+      List.all ys.toList.attach (fun ⟨ y, _ ⟩ => hasShape y shape')
+  | _, _ => false
+
+lemma hasShapeNil (x: TensorData a): ! (x.hasShape a .nil) := by
+  unfold TensorData.hasShape
+  simp
+
+lemma hasShapeRoot (xs: List a)
+  (H: (TensorData.td_root xs).hasShape a (.cons c cs)) :
+  cs = .nil ∧ List.length xs = c := by
+  revert H
+  unfold TensorData.hasShape
+  cases cs
+  all_goals simp
+
+lemma hasShapeNode
+  (H: (TensorData.td_node ys).hasShape a (.cons c cs)):
+  ys.length == c ∧ ys.all (fun y => y.hasShape a cs) := by
+  unfold TensorData.hasShape at H
+  revert H
+  simp
+  intro H
+  rw[H]
+  simp
+  clear H
+  unfold NonEmptyList.all NonEmptyList.toList
+  simp
+
+inductive TensorData.HasShape : (TensorData a) -> Shape -> Prop where
+| HasShapeRoot xs n :
+  List.length xs == n
+  -> HasShape (TensorData.td_root xs) [n]
+| HasShapeNode xs s :
+  (∀ x', x' ∈ xs.toList -> HasShape x' s)
+  -> HasShape (TensorData.td_node xs) (.cons xs.length s)
+
+theorem HashShapeOk : ∀ (x: TensorData a) (s: Shape),
+  x.HasShape a s <-> x.hasShape a s := by
+  intro x1 s1
+  constructor
+  . intro H
+    induction H with
+    | HasShapeRoot xs n H =>
+      unfold TensorData.hasShape
+      exact H
+    | HasShapeNode xs s H1 H2 =>
+      clear x1 s1 H1
+      unfold TensorData.hasShape
+      simp
+      exact H2
+  . revert x1
+    induction s1 with
+    | nil =>
+      intro x H
+      have H1 := hasShapeNil a x
+      rw[H] at H1
+      trivial
+    | cons c cs H1 =>
+      intro x
+      cases x with
+      | td_root xs =>
+        intro H1
+        have H2 := hasShapeRoot a xs H1
+        revert H2
+        rw[and_imp]
+        intro H2 H3
+        rw[H2, <- H3]
+        apply (TensorData.HasShape.HasShapeRoot xs xs.length)
+        simp
+      | td_node xs =>
+        cases xs with
+        | ncons x xs =>
+          intro H2
+          have H3 := (hasShapeNode a H2)
+          revert H3
```

Comment:
Why do you use this pattern: have ... revert ? If you revert because you want to rewrite, you could also do rw [...] at H3

Reply:
Yeah that's better.